File: app/src/main/java/com/github/overpass/gather/screen/profile/ProfileViewModel.java
package com.github.overpass.gather.screen.profile;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.core.util.Consumer;

import com.github.overpass.gather.model.usecase.image.ImageSourceUseCase;
import com.github.overpass.gather.model.usecase.userdata.ProfileUseCase;
import com.github.overpass.gather.screen.base.personal.DataViewModel;
import com.google.firebase.auth.FirebaseUser;

import javax.inject.Inject;

public class ProfileViewModel extends DataViewModel {

    private final ProfileUseCase profileUseCase;

    private boolean isEditMode = false;

    @Inject
    public ProfileViewModel(@NonNull Application application,
                            ProfileUseCase profileUseCase,
                            ImageSourceUseCase imageSourceUseCase) {
        super(application);
        setImageSourceUseCase(imageSourceUseCase);
        this.profileUseCase = profileUseCase;
    }

    public void getUserData(Consumer<FirebaseUser> onSuccess, Runnable onError) {
        profileUseCase.getUserData(onSuccess, onError);
    }

    public void signOut(Runnable onSuccess) {
        profileUseCase.signOut(onSuccess);
    }

    public void onProfileModeChanged(Consumer<Boolean> onChange) {
        isEditMode = !isEditMode;
        onChange.accept(isEditMode);
    }

    public boolean checkIfIsEditMode() {
        boolean value = isEditMode;
        if (isEditMode) {
            isEditMode = false;
        }
        return value;
    }
}

File: app/src/main/java/by/overpass/gather/ui/profile/ProfileViewModel.kt
package by.overpass.gather.ui.profile

import android.app.Application
import android.net.Uri
import androidx.core.util.Consumer
import androidx.lifecycle.MutableLiveData
import by.overpass.gather.commons.image.ChooseImageHelper
import by.overpass.gather.model.usecase.image.ImageSourceUseCase
import by.overpass.gather.model.usecase.userdata.PersonalDataUseCase
import by.overpass.gather.model.usecase.userdata.ProfileUseCase
import by.overpass.gather.ui.base.personal.DataViewModel
import com.google.firebase.auth.FirebaseUser
import com.hadilq.liveevent.LiveEvent
import javax.inject.Inject

class ProfileViewModel @Inject constructor(
        application: Application,
        chooseImageHelper: ChooseImageHelper,
        personalDataUseCase: PersonalDataUseCase,
        chosenImageData: MutableLiveData<Uri>,
        writeDeniedData: LiveEvent<Boolean>,
        readDeniedData: LiveEvent<Boolean>,
        imageSourceUseCase: ImageSourceUseCase,
        private val profileUseCase: ProfileUseCase
) : DataViewModel(
        application,
        chooseImageHelper,
        personalDataUseCase,
        chosenImageData,
        writeDeniedData,
        readDeniedData,
        imageSourceUseCase
) {

    private var isEditMode = false

    fun getUserData(onSuccess: Consumer<FirebaseUser>, onError: Runnable) {
        profileUseCase.getUserData(onSuccess, onError)
    }

    fun signOut(onSuccess: Runnable) {
        profileUseCase.signOut(onSuccess)
    }

    fun onProfileModeChanged(onChange: Consumer<Boolean>) {
        isEditMode = !isEditMode
        onChange.accept(isEditMode)
    }

    fun checkIfIsEditMode(): Boolean {
        val value = isEditMode
        if (isEditMode) {
            isEditMode = false
        }
        return value
    }
}
